embassy/src/util/signal.rs

```diff
@@ -67,3 +72,70 @@ impl<T: Send> Signal<T> {
         cortex_m::interrupt::free(|_| matches!(unsafe { &*self.state.get() }, State::Signaled(_)))
     }
 }
+
+struct NrWrap(u8);
+unsafe impl cortex_m::interrupt::Nr for NrWrap {
+    fn nr(&self) -> u8 {
+        self.0
+    }
+}
+
+pub struct InterruptFuture<'a, I: OwnedInterrupt> {
+    interrupt: &'a mut I,
+}
+
+impl<'a, I: OwnedInterrupt> Drop for InterruptFuture<'a, I> {
+    fn drop(&mut self) {
+        cortex_m::interrupt::free(|_| {
+            self.interrupt.remove_handler();
+            self.interrupt.disable();
+        });
+    }
+}
+
+impl<'a, I: OwnedInterrupt> InterruptFuture<'a, I> {
+    pub fn new(interrupt: &'a mut I) -> Self {
+        cortex_m::interrupt::free(|_| {
```

Comment:
The only thing that can race here is setting the handler, which only can race with this particular interrupt firing, not others.

If you disable before setting the handler, the critical section is not needed :)